core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/R2dbcReadJournal.scala

```diff
@@ -86,4 +91,45 @@ final class R2dbcReadJournal(system: ExtendedActorSystem, config: Config, cfgPat
       maxSlice: Int,
       offset: Offset): Source[EventEnvelope, NotUsed] =
     bySlice.liveBySlices("eventsBySlices", entityTypeHint, minSlice, maxSlice, offset)
+
+  def currentPersistenceIds(afterId: Option[String], limit: Long): Source[String, NotUsed] =
+    queryDao.persistenceIds(afterId, limit)
+
+  override def currentPersistenceIds(): Source[String, NotUsed] = {
+    import settings.querySettings.persistenceIdsBufferSize
+    def updateState(state: PersistenceIdsQueryState, pid: String): PersistenceIdsQueryState =
+      state.copy(rowCount = state.rowCount + 1, latestPid = pid)
+
+    def nextQuery(state: PersistenceIdsQueryState): (PersistenceIdsQueryState, Option[Source[String, NotUsed]]) = {
+      if (state.queryCount == 0L || state.rowCount >= persistenceIdsBufferSize) {
+        val newState = state.copy(rowCount = 0, queryCount = state.queryCount + 1)
+
+        if (state.queryCount != 0 && log.isDebugEnabled())
+          log.debug(
+            "persistenceIds query [{}] after [{}]. Found [{}] rows in previous query.",
+            state.queryCount,
+            state.latestPid,
+            state.rowCount)
+
+        newState -> Some(
+          queryDao
+            .persistenceIds(if (state.latestPid == "") None else Some(state.latestPid), persistenceIdsBufferSize))
+      } else {
+        if (log.isDebugEnabled)
+          log.debug(
+            "persistenceIds query [{}] completed. Found [{}] rows in previous query.",
+            state.queryCount,
+            state.rowCount)
+
+        state -> None
+      }
+    }
+
+    ContinuousQuery[PersistenceIdsQueryState, String](
+      initialState = PersistenceIdsQueryState(0, 0, ""),
+      updateState = updateState,
+      delayNextQuery = _ => None,
+      nextQuery = state => nextQuery(state))
+      .mapMaterializedValue(_ => NotUsed)
+  }
```

Comment:
So here ContinousQuery isn't really used for tailing but for fetching batches rather than _every_ id at once?

Reply:
yes, that is right. Same approach as for `currentEventsBySlices` and `currentEventsByPersistenceId` will be the same